Fix/8888 sync audiences and dimensions

ankitrox opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #8888 

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [ ] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.
- [x] Ensure there are no unexpected significant changes to file sizes.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

techanvil commented on assets/js/modules/analytics-4/datastore/audience-settings.js:
```diff
@@ -197,12 +181,27 @@ const baseActions = {
+			.select( MODULES_ANALYTICS_4 )
+			.getAvailableAudiences();
 
 		if ( audiences === undefined ) {
```
As per this [comment](https://github.com/google/site-kit-wp/pull/10264/files#r1995803710), this condition is incorrect, we should retain the check for `null` as in the original `*getAvailableAudiences()`:

techanvil replied:
On a somewhat related note, I've noticed this retrieval of `availableAudiences` hasn't been updated in this PR:

https://github.com/google/site-kit-wp/blob/c1c5f2b434789aa1d9c541159ff5cca3deb8110f/assets/js/modules/analytics-4/index.js#L198-L199

techanvil reviewed (changes requested):
Hi @ankitrox, good progress on this, but it needs a few amendments.

Please see my comments, and in addition:

- Merge in `develop` and fix the conflicts. The Audience Segmentation feature isn't fully functional without this.
- Add the DB migration script that will be required.
- Add a point to the QAB, a general smoke test should be performed for the AS feature in addition to the QA:Eng steps that have been outlined.

Thanks!

ankitrox commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/ChangeGroupsLink.test.js:
```diff
@@ -36,27 +36,15 @@ mockTrackEvent.mockImplementation( () => Promise.resolve() );
-
 	beforeEach( () => {
 		registry = createTestRegistry();
+		provideUserAuthentication( registry );
```
This doesn't seem to be required, remove it.

ankitrox commented on includes/Modules/Analytics_4.php:
```diff
@@ -262,24 +262,17 @@ public function register() {
 
-					$this->resource_data_availability_date->reset_all_resource_dates( $available_audience_names, $old_value['propertyID'] );
-				}
+				$this->resource_data_availability_date->reset_all_resource_dates( $available_audience_names );
```
Hi @techanvil,
This call is already inside `$this->audience_settings->on_change`, thus it will be called every time when Audience Settings is changed.

Please let me know if I misunderstood it.

techanvil replied:
Hey @ankitrox, that's right, but it's a change from the current behaviour where `reset_all_resource_dates()` is only invoked when the property or measurement ID has changed. To illustrate, here's the current version of `Analytics_4.php` on the left vs the new on the right:

![image](https://github.com/user-attachments/assets/1adeae7f-ba2a-446d-9bc1-84e3490debf5)

ankitrox commented:
Thank you so much for reviewing the PR thoroughly and adding your valuable suggestions @techanvil .

I have made the necessary and suggested changes along with the update in QAB and resolved merge conflicts.

Assigning it to you to review the changes made in the PR.

Thanks again!

techanvil commented on includes/Core/Util/Migration_1_150_0.php:
This migration class needs to be instantiated and registered in `Plugin.php`.

techanvil commented on tests/phpunit/integration/Modules/Analytics_4/Resource_Data_Availability_DateTest.php:
```diff
@@ -47,16 +53,22 @@ public function set_up() {
+			array(
 				'availableAudiences' => array(
-					array(
+					$this->test_audience_1 => array(
```
This doesn't look correct, `availableAudiences` should be a list of objects, not an associative array.

techanvil commented on tests/phpunit/integration/Modules/Analytics_4/Resource_Data_Availability_DateTest.php:
```diff
@@ -91,6 +103,8 @@ public function test_reset_resource_date() {
 	}
 
 	public function test_get_all_resource_dates() {
+		$this->enable_feature( 'audienceSegmentation' );
```
The `audienceSegmentation` feature flag has been removed and doesn't need enabling here.

techanvil commented on tests/phpunit/integration/Modules/Analytics_4/Resource_Data_Availability_DateTest.php:
```diff
@@ -153,6 +167,8 @@ public function test_reset_all_resource_dates() {
 	}
 
 	public function test_is_valid_resource_slug() {
+		$this->enable_feature( 'audienceSegmentation' );
```
As above.

techanvil commented on includes/Modules/Analytics_4.php:
```diff
@@ -348,9 +356,15 @@ function ( $audience ) {
-					$new_value['audienceSegmentationSetupCompletedBy'] = null;
+					$new_value['availableCustomDimensions'] = null;
+
+					$this->audience_settings->set(
+						array(
+							'availableAudiences' => null,
+							'availableAudiencesLastSyncedAt' => 0,
+							'audienceSegmentationSetupCompletedBy' => null,
+						)
+					);
```
Please remove this call to `$this->audience_settings->set()` which is no longer needed in the `pre_update_` hook.

techanvil commented on includes/Modules/Analytics_4.php:
```diff
@@ -307,6 +290,27 @@ function ( $audience ) {
+					$this->reset_data_available();
+					$this->custom_dimensions_data_available->reset_data_available();
+
+					$available_audiences = $old_value['availableAudiences'] ?? array();
```
`$available_audiences` will need to be retrieved from `$this->audience_settings` here, as it's no longer in the module settings.

techanvil reviewed (changes requested):
Thanks for the update @ankitrox, however some further changes are needed on this one.

ankitrox commented:
Thank you @techanvil for reviewing the change again and adding the feedback.

I have added back the `getAvailableAudiences` resolver as mentioned and fixed the tests and stories accordingly.

Can you please review these changes one more time to confirm if these are looking fine now?

Thanks.

techanvil commented on assets/js/modules/analytics-4/datastore/audiences.test.js:
```diff
@@ -2407,7 +2395,7 @@ describe( 'modules/analytics-4 audiences', () => {
 				registry
 					.dispatch( MODULES_ANALYTICS_4 )
-					.receiveGetSettings( {} );
+					.startResolution( 'getAvailableAudiences', [] );
```
We can just change this to use `receiveGetAudienceSettings()`. Please check the other cases where `startResolution()` has been applied, this might have been due to the check for `undefined` in `*getAvailableAudiences()` and it seems likely they can be changed to use `receiveGetAudienceSettings()` which is a bit simpler to understand.

```suggestion
					.receiveGetAudienceSettings( {} );
```

techanvil commented on assets/js/modules/analytics-4/datastore/audiences.test.js:
```diff
@@ -2466,10 +2470,24 @@ describe( 'modules/analytics-4 audiences', () => {
 			it( 'should return `true` if the audience is a user audience', () => {
-				registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
-					availableAudiences: availableAudiencesFixture,
+				fetchMock.postOnce( syncAvailableAudiencesEndpoint, {
+					body: availableAudiencesFixture,
+					status: 200,
 				} );
 
+				registry.dispatch( CORE_USER ).receiveGetUserAudienceSettings( {
+					configuredAudiences: [
+						'properties/12345/audiences/1',
+						'properties/12345/audiences/2',
+					],
+				} );
```
As above:

```suggestion
```

techanvil commented on assets/js/modules/analytics-4/datastore/audiences.test.js:
```diff
@@ -2779,12 +2816,24 @@ describe( 'modules/analytics-4 audiences', () => {
 					.dispatch( MODULES_ANALYTICS_4 )
 					.receiveIsGatheringData( false );
+
+				fetchMock.postOnce( syncAvailableAudiencesEndpoint, {
+					body: availableAudiencesFixture,
+					status: 200,
+				} );
+
+				registry.dispatch( CORE_USER ).receiveGetUserAudienceSettings( {
+					configuredAudiences: [
+						'properties/12345/audiences/1',
+						'properties/12345/audiences/2',
+					],
+				} );
```
As above:

```suggestion
```

techanvil commented on assets/js/modules/analytics-4/datastore/audiences.test.js:
```diff
@@ -1067,6 +1085,14 @@ describe( 'modules/analytics-4 audiences', () => {
+					.dispatch( MODULES_ANALYTICS_4 )
+					.receiveGetAudienceSettings( {
+						availableAudiences,
+						availableAudiencesLastSyncedAt:
+							( Date.now() - 1000 ) / 1000,
```
We don't need to set the timestamp here:

```suggestion
```

techanvil commented on assets/js/modules/analytics-4/datastore/audiences.test.js:
```diff
@@ -1292,6 +1323,11 @@ describe( 'modules/analytics-4 audiences', () => {
 
 					muteFetch( expirableItemEndpoint );
 
+					fetchMock.postOnce( syncAvailableAudiencesEndpoint, {
+						body: JSON.stringify( availableAudiencesFixture ),
+						status: 200,
+					} );
```
This isn't needed:

```suggestion
```

techanvil commented on assets/js/modules/analytics-4/datastore/audience-settings.js:
```diff
@@ -257,6 +271,10 @@ const baseSelectors = {
 		return state.audienceSettings;
 	},
+
+	getAvailableAudiencesLastSyncedAt( state ) {
+		return state.audienceSettings?.availableAudiencesLastSyncedAt;
+	},
```
Hi @ankitrox, I've just spotted another thing that we need to fix here, that should have been addressed in a previous issue.

All these individual audience setting selectors need to call `getAudienceSettings()` in order to invoke its resolver. So, this one should be updated as follows:

```suggestion
	/**
	 * Gets the last time the available audiences were synced.
	 *
	 * @since n.e.x.t
	 *
	 * @param {Object} state Data store's state.
	 * @return {(number|undefined)} Last time the available audiences were synced, or `undefined` if not loaded.
	 */
	getAvailableAudiencesLastSyncedAt: createRegistrySelector(
		( select ) => () => {
			const audienceSettings =
				select( MODULES_ANALYTICS_4 ).getAudienceSettings() || {};

			return audienceSettings.availableAudiencesLastSyncedAt;
		}
	),
```

Please can you apply this to the other selectors too? Also, don't forget the JSDoc :)

techanvil reviewed (changes requested):
Hi @ankitrox, this needs quite a substantial cleanup, mostly due to test fixes that are no longer needed. There are also a couple of important functional aspects that were missed. Please take a look.

ankitrox commented:
Thanks @techanvil for adding the suggestions for the cleanup. Those have been addressed and moved to CR again.

techanvil reviewed (changes requested):
Thanks @ankitrox, it looks like this is nearly over the line. I noticed a handful of minor JS issues that still needed tidying up, which I pushed a commit for myself to save time.

However, there are a couple of outstanding PHPUnit test issues which I'm sending it back to you to address. Please take a look.

ankitrox commented:
Thank you so much @techanvil for helping here with phpunit issue. Rest of the things have already been addressed, so sending your way for another review.

techanvil commented on includes/Modules/Analytics_4/Audience_Settings.php:
```diff
@@ -99,6 +99,19 @@ public function merge( $settings ) {
 		return $updated_settings;
 	}
 
+	/**
+	 * Gets the settings.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array The settings.
+	 */
+	public function get() {
+		$option = $this->options->get( self::OPTION );
+
+		return false === $option ? $this->get_default() : $option;
+	}
+
```
This shouldn't be needed. As per the JSDoc for `Setting::get()`, the registered default value will be returned if the option isn't set.

https://github.com/google/site-kit-wp/blob/835ece6a20f4fe25d32c90d8757cf82c2f49c88f/includes/Core/Storage/Setting.php#L103-L113

It looks like this was added to fix the `test_module_level_audience_settings_reset__on_deactivation()` test, but the correct fix for that test would be to add `$this->analytics->register()` at the top of it, in order to register the setting so its default will be returned.

techanvil commented on includes/Modules/Analytics_4/Audience_Settings.php:
```diff
@@ -108,7 +121,7 @@ public function merge( $settings ) {
 	 * @return array The sanitized settings.
 	 */
-	private function sanitize( $option ) {
+	private function sanitize( $option = array() ) {
```
Why has this default value been added? It's not something that's typically needed for our sanitize callbacks...

techanvil reviewed (changes requested):
Thanks @ankitrox! Looks like this is on the verge of being mergeable, but I have left a couple more comments, please take a look.

techanvil reviewed (approved):
LGTM. Thanks for all you work getting this one over the line @ankitrox!

:white_check_mark: 

Note that I've confirmed the failing E2E tests are unrelated to the changes in this PR.
